### PR TITLE
Support cursor-relative operations (and add intellisense to ReadCursor)

### PR DESCRIPTION
**Pull request by mgravell**
Allows:

```
var cursor2 = cursor1 + 17;
```

and

```
int delta = cursor2 - cursor1; // 17
```

### CODE REVIEW

**Comment by Drawaes** [comment]
Good work, very useful "book marking" by taking a cursor is much more useful now!

**Comment by mgravell** [comment]
Note refactored this to use `Seek`, but IMO the deleted implementation in `+` (see commit history) is cleaner and easier to grok; _might_ be worth considering resurrecting that and making it the master? Although would need to respect the existing Seek semantic of stopping at the end rather than throwing

File: src/Channels/ReadCursor.cs [hunk 1]
```diff
@@ -272,5 +298,68 @@ public override int GetHashCode()
...
+        /// <summary>
+        /// Calcualte the offset between two cursors
+        /// </summary>
+        /// <remarks>The first operand must be equal or greater than the second operand -
+        /// meaning: the expected result must be non-negative</remarks>
+        public static int operator -(ReadCursor later, ReadCursor earlier)
+        {
+            var segment = earlier.Segment;
+            int index = earlier.Index, delta = 0;
+            while(segment != null)
```
Comment (davidfowl) [inline src/Channels/ReadCursor.cs:333]: nit: formatting (ctrl + k, ctrl + d)

**Review by davidfowl — CHANGES_REQUESTED** [review]
Do you think these operations are useful when you can't really get cursors outside of slices?

**Comment by mgravell** [comment]
Yes, in particular subtraction for asking about lengths between two points
("how much got written?"). The addition is semantically cleaner than
`Slice(...).Start` if you need to talk about pos+n.

File: test/Channels.Tests/ReadCursorFacts.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                Assert.Throws<ArgumentOutOfRangeException>(() => rb.Start + (-1));
+                Assert.Throws<ArgumentOutOfRangeException>(() => rb.End + (-1));
+                Assert.Throws<ArgumentOutOfRangeException>(() => rb.End + 1);
+
+                
+                for (int i = 0; i <= 90; i++)
+                {
+                    try
+                    {
+                        if (i == 30) System.Diagnostics.Debugger.Break();
```
Comment (davidfowl) [inline test/Channels.Tests/ReadCursorFacts.cs:52]: Remove debugger.break

File: test/Channels.Tests/ReadCursorFacts.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                        var viaOp = rb.Start + i;
+                        var viaSlice = rb.Slice(i).Start;
+                        Assert.Equal(viaSlice, viaOp);
+
+                        Assert.Equal(i, viaOp - rb.Start);
+                        Assert.Equal(90 - i, rb.End - viaOp);
+                    }
+                    catch(Exception ex)
+                    {
+                        throw new Exception($"Failed for {nameof(i)}={i}: {ex.Message}", ex);
```
Comment (davidfowl) [inline test/Channels.Tests/ReadCursorFacts.cs:62]: wut?

**Review by davidfowl — CHANGES_REQUESTED** [review]

**Pull request closed without merging**